source/carton-bindings-ocaml/src/export.rs

```diff
@@ -0,0 +1,33 @@
+use crate::*;
+
+#[ocaml_rust::bridge]
+mod ffi {
+    type FFICarton = CustomConst<Carton>;
+    type FFICartonError = Custom<CartonError>;
+
+    #[derive(Clone, Debug)]
+    pub enum Tensor {
+        Float(Vec<f32>),
+        Double(Vec<f64>),
+        I8(Vec<i64>),
+        I16(Vec<i64>),
+        I32(Vec<i64>),
+        I64(Vec<i64>),
+        U8(Vec<i64>),
+        U16(Vec<i64>),
+        U32(Vec<i64>),
+        U64(Vec<i64>),
+        String(Vec<String>),
+    }
```

Comment:
Here's the PR: #167

Reply:
This should actually help a lot implementing things.

The reason I am copying and such is that ocaml only supports `Float`, `Double`, `int32`, and `int64`, so we cant share by reference by ocaml could not understand some of the smaller ints. There are libraries to add support for other types but I believe if we wanted to make it have all the types in rust we would have have to fork `ocaml-rust`.

Reply:
I will also update my enum to have a i32 for the lower values if there is no way to not copy the data.

Reply:
That makes sense. As I mentioned in my other comment, it's been a long time since I've written OCaml so I don't have a whole lot of context. Is there a popular ndarray library for OCaml that we could use in the interface?

For example:
* Python has Numpy,
* Rust has the `ndarray` crate
* JavaScript has a few usable ones on top of `TypedArray`s
* C lets us pass pointers to data (along with a deleter function to free the data when it's no longer needed)

Reply:
Let me see if I understand what these libraries accomplish. Is the goal to have a chunk of memory with a fixed length, and the "basic" types of a language (the ones specified in the tensor enum)?

I believe in Ocaml the `Base.Array` is a vector (I believe what its called when its not a linked list) so all the data will be in the same block of memory. To add support for the different number types we can use [stdint](https://github.com/andrenth/ocaml-stdint) (will give us support for the rest of the ints needed) then all the data would be the same as a rust vector.

If I am misunderstanding the reason for the need of these types, then there does look to be [np](https://github.com/lehy/ocaml-sklearn) for Ocaml that give support for matrixes and multi dementinal arrays.